iphone/CoreApi/CoreApi/Logger/Logger.mm

```diff
@@ -1,32 +1,230 @@
 #import "Logger.h"
-#import "base/logging.hpp"
+#import <OSLog/OSLog.h>
+
+#include "base/logging.hpp"
+#include "coding/zip_creator.hpp"
 
 @interface Logger ()
+
+@property (nullable, nonatomic) NSFileHandle * fileHandle;
+@property (nonnull, nonatomic) os_log_t osLogger;
+@property (class, nonatomic) BOOL fileLoggingEnabled;
+@property (class, readonly, nonatomic) dispatch_queue_t fileLoggingQueue;
+
++ (Logger *)logger;
++ (void)enableFileLogging;
++ (void)disableFileLogging;
++ (void)logMessageWithLevel:(base::LogLevel)level src:(base::SrcPoint const &)src message:(std::string const &)message;
++ (NSURL *)getZippedLogFile:(NSString *)logFilePath;
++ (void)removeFileAtPath:(NSString *)filePath;
 + (base::LogLevel)baseLevel:(LogLevel)level;
+
 @end
 
+// Subsystem and category are used for the OSLog.
+NSString * const kLoggerSubsystem = [[NSBundle mainBundle] bundleIdentifier];
+NSString * const kLoggerCategory = @"OM";
+NSString * const kLogFileName = @"log.txt";
+NSString * const kZipLogFileExtension = @"zip";
+NSString * const kLogFilePath = [[NSFileManager.defaultManager temporaryDirectory] URLByAppendingPathComponent:kLogFileName].path;
+NSUInteger const kMaxLogFileSize = 1024 * 1024 * 100; // 100 MB;
+
 @implementation Logger
 
-+ (void)log:(LogLevel)level message:(NSString*)message {
-  LOG_SHORT([Logger baseLevel:level], (message.UTF8String));
+static BOOL _isFileLoggingEnabled = NO;
+
++ (void)initialize
+{
+  if (self == [Logger class]) {
+    SetLogMessageFn(&LogMessage);
+    SetAssertFunction(&AssertMessage);
+  }
+}
+
++ (Logger *)logger {
+  static Logger * logger = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    logger = [[self alloc] init];
+  });
+  return logger;
+}
+
++ (dispatch_queue_t)fileLoggingQueue {
+  static dispatch_queue_t fileLoggingQueue = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    dispatch_queue_attr_t attributes = dispatch_queue_attr_make_with_qos_class(DISPATCH_QUEUE_SERIAL, QOS_CLASS_UTILITY, 0);
+    fileLoggingQueue = dispatch_queue_create("app.organicmaps.fileLoggingQueue", attributes);
+  });
+  return fileLoggingQueue;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _osLogger = os_log_create(kLoggerSubsystem.UTF8String, kLoggerCategory.UTF8String);
+  }
+  return self;
+}
+
+// MARK: - Public
+
++ (void)setFileLoggingEnabled:(BOOL)fileLoggingEnabled {
+  fileLoggingEnabled ? [self enableFileLogging] : [self disableFileLogging];
+}
+
++ (BOOL)fileLoggingEnabled {
+  return _isFileLoggingEnabled;
```

Comment:
Can this BOOL be avoided and NSUserDefaults used instead to have only one source of truth and no need to sync them?

Reply:
This BOOL is already in the NSUserDefaults handeled in the MWMSettings class.
Its better to store all the settings at one place. The MWMSettings - is the source of truth for the all app settings.